freqtrade/freqai/freqai_interface.py

```diff
@@ -0,0 +1,572 @@
+# import contextlib
+import copy
+import datetime
+import gc
+import logging
+import shutil
+import threading
+import time
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Any, Dict, Tuple
+
+import numpy as np
+import pandas as pd
+from numpy.typing import ArrayLike
+from pandas import DataFrame
+
+from freqtrade.configuration import TimeRange
+from freqtrade.enums import RunMode
+from freqtrade.exceptions import OperationalException
+from freqtrade.freqai.data_drawer import FreqaiDataDrawer
+from freqtrade.freqai.data_kitchen import FreqaiDataKitchen
+from freqtrade.strategy.interface import IStrategy
+
+
+pd.options.mode.chained_assignment = None
+logger = logging.getLogger(__name__)
+
+
+def threaded(fn):
+    def wrapper(*args, **kwargs):
+        threading.Thread(target=fn, args=args, kwargs=kwargs).start()
+
+    return wrapper
+
+
+class IFreqaiModel(ABC):
+    """
+    Class containing all tools for training and prediction in the strategy.
+    Base*PredictionModels inherit from this class.
+
+    Record of contribution:
+    FreqAI was developed by a group of individuals who all contributed specific skillsets to the
+    project.
+
+    Conception and software development:
+    Robert Caulk @robcaulk
+
+    Theoretical brainstorming:
+    Elin Törnquist @thorntwig
+
+    Code review, software architecture brainstorming:
+    @xmatthias
+
+    Beta testing and bug reporting:
+    @bloodhunter4rc, Salah Lamkadem @ikonx, @ken11o2, @longyu, @paranoidandy, @smidelis, @smarm
+    Juha Nykänen @suikula, Wagner Costa @wagnercosta
+    """
+
+    def __init__(self, config: Dict[str, Any]) -> None:
+
+        self.config = config
+        self.assert_config(self.config)
+        self.freqai_info = config["freqai"]
+        self.data_split_parameters = config.get("freqai", {}).get("data_split_parameters")
+        self.model_training_parameters = config.get("freqai", {}).get("model_training_parameters")
+        self.feature_parameters = config.get("freqai", {}).get("feature_parameters")
+        self.retrain = False
+        self.first = True
+        self.set_full_path()
+        self.follow_mode = self.freqai_info.get("follow_mode", False)
+        self.dd = FreqaiDataDrawer(Path(self.full_path), self.config, self.follow_mode)
+        self.lock = threading.Lock()
+        self.identifier = self.freqai_info.get("identifier", "no_id_provided")
+        self.scanning = False
+        self.keras = self.freqai_info.get("keras", False)
+        if self.keras and self.freqai_info.get("feature_parameters", {}).get("DI_threshold", 0):
+            self.freqai_info["feature_parameters"]["DI_threshold"] = 0
+            logger.warning("DI threshold is not configured for Keras models yet. Deactivating.")
+        self.CONV_WIDTH = self.freqai_info.get("conv_width", 2)
```

Comment:
This parameter (as well as keras) is neither documented, nor part of the schema.

The 2nd part is not necessarily a problem (but means you must type-check it before usage) - but yet i'd refer that to the schema.